cas/tree.go

```diff
@@ -0,0 +1,354 @@
+		if err != nil {
+			return fmt.Errorf("treestore: failed to open tree store directory: %v", err)
+		}
+		defer f.Close()
+		err = f.Sync()
+		if err != nil {
+			return fmt.Errorf("treestore: failed to sync tree store directory: %v", err)
+		}
+	}
+	return os.RemoveAll(treepath)
+}
+
+// IsRendered checks if the tree store is fully rendered
+func (ts *TreeStore) IsRendered(key string) (bool, error) {
+	// if the "rendered" flag file exists, assume that the store is already
+	// fully rendered.
+	treepath := filepath.Join(ts.path, key)
+	_, err := os.Stat(filepath.Join(treepath, renderedfilename))
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// GetPath returns the absolute path of the treestore for the specified key.
+// It doesn't ensure that the path exists and is fully rendered. This should
+// be done calling IsRendered()
+func (ts *TreeStore) GetPath(key string) string {
+	return filepath.Join(ts.path, key)
+}
+
+// GetRootFS returns the absolute path of the rootfs for the specified key.
+// It doesn't ensure that the rootfs exists and is fully rendered. This should
+// be done calling IsRendered()
+func (ts *TreeStore) GetRootFS(key string) string {
+	return filepath.Join(ts.GetPath(key), "rootfs")
+}
+
+// TreeStore calculates an hash of the rendered ACI. It uses the same functions
+// used to create a tar but instead of writing the full archive is just
+// computes the sha512 sum of the file infos and contents.
+func (ts *TreeStore) Hash(key string) (string, error) {
+	treepath := filepath.Join(ts.path, key)
+
+	hash := sha512.New()
+	iw := NewHashWriter(hash)
+	err := filepath.Walk(treepath, buildWalker(treepath, iw))
+	if err != nil {
+		return "", fmt.Errorf("treestore: error walking rootfs: %v", err)
+	}
+
+	hashstring := hashToKey(hash)
+
+	return hashstring, nil
+}
+
+// Check calculates the actual rendered ACI's hash and verifies that it matches
+// the saved value.
+func (ts *TreeStore) Check(key string) error {
+	treepath := filepath.Join(ts.path, key)
+	hash, err := ioutil.ReadFile(filepath.Join(treepath, hashfilename))
+	if err != nil {
+		return fmt.Errorf("treestore: cannot read hash file: %v", err)
+	}
+	curhash, err := ts.Hash(key)
+	if err != nil {
+		return fmt.Errorf("treestore: cannot calculate tree hash: %v", err)
+	}
+	if curhash != string(hash) {
+		return fmt.Errorf("treestore: wrong tree hash: %s, expected: %s", curhash, hash)
+	}
+	return nil
+}
+
+type xattr struct {
+	Name  string
+	Value string
+}
+
+// Like tar Header but, to keep json output reproducible:
+// * Xattrs as a slice
+// * Skip Uname and Gname
+// TODO. Should ModTime/AccessTime/ChangeTime be saved? For validation its
+// probably enough to hash the file contents and the other infos and avoid
+// problems due to them changing.
+// TODO(sgotti) Is it possible that json output will change between go
+// versions? Use another or our own Marshaller?
+type fileInfo struct {
```

Comment:
@vbatts this is slightly more complicated as it's a rendered aci (files can come from multiple ACIs). I hadn't time to look in tar-split but perhaps, the information's can be extracted at rendering time by pkg/aci/render.renderImage instead of computing the hash after the rendering (as done in this patch).